plutus-tx/src/Language/PlutusTx/Ratio.hs

```diff
@@ -0,0 +1,205 @@
+toGHC :: Rational -> Ratio.Rational
+toGHC (n :% d) = n Ratio.:% d
+
+{-# INLINABLE numerator #-}
+-- | Extract the numerator of the ratio in reduced form: the numerator and denominator have no common factor and the denominator is positive.
+numerator :: Ratio a -> a
+numerator (n :% _) = n
+
+{-# INLINABLE denominator #-}
+-- | Extract the denominator of the ratio in reduced form: the numerator and denominator have no common factor and the denominator is positive.
+denominator :: Ratio a -> a
+denominator (_ :% d) = d
+
+{-# INLINABLE gcd #-}
+-- From GHC.Real
+-- | @'gcd' x y@ is the non-negative factor of both @x@ and @y@ of which
+-- every common factor of @x@ and @y@ is also a factor; for example
+-- @'gcd' 4 2 = 2@, @'gcd' (-4) 6 = 2@, @'gcd' 0 4@ = @4@. @'gcd' 0 0@ = @0@.
+gcd :: Integer -> Integer -> Integer
+gcd a b = gcd' (abs a) (abs b) where
+    gcd' a' b'
+        | b' P.== P.zero = a'
+        | True           = gcd' b' (a' `Builtins.remainderInteger` b')
+
+{-# INLINABLE truncate #-}
+-- | truncate @x@ returns the integer nearest @x@ between zero and @x@
+truncate :: Ratio Integer -> Integer
+truncate r = let (m, _ :: Rational) = properFraction r in m
+
+{-# INLINABLE properFraction #-}
+-- From GHC.Real
+-- | The function 'properFraction' takes a real fractional number @x@
+-- and returns a pair @(n,f)@ such that @x = n+f@, and:
+--
+-- * @n@ is an integral number with the same sign as @x@; and
+--
+-- * @f@ is a fraction with the same type and sign as @x@,
+--   and with absolute value less than @1@.
+--
+-- The default definitions of the 'ceiling', 'floor', 'truncate'
+-- and 'round' functions are in terms of 'properFraction'.
+properFraction :: Ratio Integer -> (Integer, Ratio Integer)
+properFraction (n :% d) = (q, r :% d) where (q, r) = quotRem n d
+
+{-# INLINABLE quot #-}
+quot :: Integer -> Integer -> Integer
+quot x y = let (q, _ :: Integer) = quotRem x y in q
+
+{-# INLINABLE quotRem #-}
+-- | simultaneous quot and rem
+quotRem :: Integer -> Integer -> (Integer, Integer)
+quotRem x y = 
+    -- `quot` is integer division truncated towards 0, while the result of `div`
+    -- is truncated toward negative inf.
+    -- Cf. https://stackoverflow.com/questions/339719/when-is-the-difference-between-quotrem-and-divmod-useful
+    -- So we need to add 1 in some cases!
+    let d   = x `Builtins.divideInteger` y
+        rem = x `Builtins.remainderInteger` y
+    in if d P.< P.zero P.&& rem P./= P.zero
+       then (d P.+ P.one, rem)
+       else (d, rem)
+
+{-# INLINABLE half #-}
+-- | 0.5
+half :: Ratio Integer
+half = 1 :% 2
+
+{-# INLINABLE reduce #-}
+-- | From GHC.Real
+-- | 'reduce' is a subsidiary function used only in this module.
+-- It normalises a ratio by dividing both numerator and denominator by
+-- their greatest common divisor.
+reduce :: Integer -> Integer -> Ratio Integer
+reduce x y
+    | y P.== 0 = Builtins.error ()
+    | True     = 
+        let d = gcd x y in
+        (x `quot` d) :% (y `quot` d)
+
```

Comment:
I'm a bit dubious about  this as well.  `reduce 144 8` is 18 :% 1, but `reduce (-144) (-8)` is (-18) :% (-1)`.  Maybe that doesn't matter for the way it's used here?

Reply:
Note that `gcd a b = gcd' (abs a) (abs b)` so `reduce 144 8 = reduce (-144) (-8) = 18 :% 1`. This is all copied verbatim from [`GHC.Real`](https://hackage.haskell.org/package/base-4.12.0.0/docs/src/GHC.Real.html#gcd)

I know there are probably other ways to implement this but I wanted to stay as close to GHC's implementation as possible (not reinventing the wheel etc.)